docs/chained-sessions.md

```diff
@@ -0,0 +1,161 @@
+
+This page documents how this feature works.
+
+## The `nextSession` URL
+
+An [extended session request](session-requests.md#extra-parameters) may now contain a `nextSession` object which must contain a `url` field. If so, then at the end of the session (i.e., after the user has agreed to perform the session in the session screen of the IRMA app), the IRMA server will POST the [session result](https://pkg.go.dev/github.com/privacybydesign/irmago/server#SessionResult) in JSON (as returned by the [`/result` endpoint](api-irma-server#get-session-requestortoken-result)) to that `url`. The server at that `url` must then respond with one of the following:
+
+* a new (extended) session request in JSON, which may depend on the received session result;
+* HTTP 204, signifying that there is no next session to be performed. In this case, no further sessions take place and the flow stops normally.
+
+In the first case, the IRMA server will then start a new session using that session request and pass it to the IRMA app, which will show the corresponding session screen. Thus, the app user moves from the first session screen immediately to the next one.
+
+The session request returned by the server at the `nextSession` URL may itself contain a `nextSession` object (possibly but not necessarily referring to the same server). If so, then yet another session will be started using the same mechanism, after the one specified by the current session request. In this fashion, a session chain can consist of any number of sessions (although ideally it is kept as short as possible for optimal UX; in most cases two sessions will suffice).
+
+## Signing POSTed session results
+
+If a [JWT private key is installed in the IRMA server](irma-server.md#signed-jwt-session-results), then instead of POSTing plain JSON session results (as returned by the [`/result` endpoint](api-irma-server#get-session-requestortoken-result)) to the `nextSession` URL, the server will POST a session result JWT signed with the private key (as returned by the [`/result-jwt` endpoint](api-irma-server.md#get-session-requestortoken-result-jwt)) to the `nextSession` URL. The server at that URL can verify the JWT using the corresponding public key to authenticate the request as coming from the expected IRMA server.
+
+If a JWT private key is not installed, then the boolean `--allow-unsigned-callbacks` option must be passed to the IRMA server before chained sessions may be used, to explicitly enable POSTing unsigned session results. Otherwise, the server will reject session requests containing a `nextSession` object.
+
+> If no JWT private key is installed, then the `nextSession` URL should either not be publically reachable, or it should be kept secret with TLS enabled (which it should anyway as personal data will be POSTed to it). Otherwise there is no way for the server at the `nextSession` URL to distinguish POSTs from your IRMA server from POSTs made by anyone else.
```

Comment:
With having a secret URL, I meant having a URL such as `http://example.com/myNextSessionEndpoint/cX5aTins5kEZpjDpfYcN`. Why is that security by obscurity? Since the path of the URL is sent to the server only after the TLS connection has been established, that should be exactly as safe as instead using an authentication token in a HTTP header, which is standard practice. Additionally, if you do it that way I don't see the need of having a domain or IP address whitelist: outsiders will not know that URL, they cannot learn it by observing any TLS sessions of that server, so they will be unable to reach that endpoint.

But perhaps that sentence makes it insufficiently clear that I meant having a secret token in the URL?

Reply:
Ah okay that makes sense. Then it is maybe nice to explicitly mention that they have to use some kind of security token / uniqueness / ... in their public URL. I don't know a good word for it.

Reply:
Done. NB: I based that sentence on a similar sentence from the `irma server` documentation, so I changed it there too.